java/unittests/infodynamics/measures/continuous/kraskov/DualTotalCorrelationCalculatorKraskovTester.java

```diff
@@ -0,0 +1,143 @@
+import infodynamics.utils.MatrixUtils;
+import infodynamics.utils.RandomGenerator;
+import junit.framework.TestCase;
+
+public class DualTotalCorrelationCalculatorKraskovTester extends TestCase {
+
+  /**
+   * For two variables, DTC is equal to mutual information.
+   */
+	public void testTwoVariables() throws Exception {
+
+    double[][] data;
+		ArrayFileReader afr = new ArrayFileReader("demos/data/2randomCols-1.txt");
+		data = afr.getDouble2DMatrix();
+
+		DualTotalCorrelationCalculatorKraskov dtcCalc = new DualTotalCorrelationCalculatorKraskov();
+    dtcCalc.setProperty("NOISE_LEVEL_TO_ADD", "0");
+    dtcCalc.initialise(2);
+    dtcCalc.setObservations(data);
+    double dtc = dtcCalc.computeAverageLocalOfObservations();
+
+
+		MutualInfoCalculatorMultiVariateKraskov1 miCalc = new MutualInfoCalculatorMultiVariateKraskov1();
+    miCalc.setProperty("NOISE_LEVEL_TO_ADD", "0");
+    miCalc.initialise(1,1);
+    miCalc.setObservations(MatrixUtils.selectColumn(data, 0),
+                           MatrixUtils.selectColumn(data, 1));
+    double mi  = miCalc.computeAverageLocalOfObservations();
+
+    assertEquals(dtc, mi, 1e-6);
+	}
+
+  /**
+   * Compare against the values obtained by the Gaussian DTC calculator when the data
+   * is actually Gaussian.
+   */
+  public void testCompareWithGaussian() throws Exception {
+
+    int N = 100000, D = 3;
+		RandomGenerator rg = new RandomGenerator();
+    rg.setSeed(1);
+    double[][] data = rg.generateNormalData(N, D, 0, 1);
+    for (int i = 0; i < N; i++) {
+      data[i][2] = data[i][2] + 0.1*data[i][0];
+      data[i][1] = data[i][1] + 0.5*data[i][0];
+    }
+
+		DualTotalCorrelationCalculatorKraskov dtcCalc_ksg = new DualTotalCorrelationCalculatorKraskov();
+    dtcCalc_ksg.setProperty("NOISE_LEVEL_TO_ADD", "0");
+    dtcCalc_ksg.initialise(3);
+    dtcCalc_ksg.setObservations(data);
+    double dtc_ksg = dtcCalc_ksg.computeAverageLocalOfObservations();
+
+		DualTotalCorrelationCalculatorGaussian dtcCalc_gau = new DualTotalCorrelationCalculatorGaussian();
+    dtcCalc_gau.initialise(3);
+    dtcCalc_gau.setObservations(data);
+    double dtc_gau = dtcCalc_gau.computeAverageLocalOfObservations();
+
+    assertEquals(dtc_ksg, dtc_gau, 0.001);
+
+  }
+
+	/**
+	 * Confirm that the local values average correctly back to the average value
+	 * 
+	 */
+	public void testLocalsAverageCorrectly() throws Exception {
+		
+    int dimensions = 4;
+    int timeSteps = 1000;
+		DualTotalCorrelationCalculatorKraskov dtcCalc = new DualTotalCorrelationCalculatorKraskov();
+		dtcCalc.initialise(dimensions);
+		
+		// generate some random data
+		RandomGenerator rg = new RandomGenerator();
+		double[][] data = rg.generateNormalData(timeSteps, dimensions,
+				0, 1);
+		
+		dtcCalc.setObservations(data);
+		
+		double dtc = dtcCalc.computeAverageLocalOfObservations();
+		double[] dtcLocal = dtcCalc.computeLocalOfPreviousObservations();
+		
+		System.out.printf("Average was %.5f\n", dtc);
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.